### PR TITLE
Discover trainable parameters using reverse DFS from loss node

### PR DESCRIPTION
**Pull request by suffiank**
**Description**: 
Discover trainable parameters using reverse DFS from loss node, omitting recursion along untrainable inputs.

_Why is this change required? What problem does it solve?_
The existing discovery process for trainable parameters assumes all initializers are trainable unless they they match a particular node op type and input index. This can fail to detect that an otherwise trainable parameter is blocked by a non-trainable parameter. For example, consider a => b => c. If c is trainable and b is non trainable, then this implies a is also not trainable.

The user can inspect the found list of trainable parameters with `TrainingSession::GetWeights()`

### CODE REVIEW

File: onnxruntime/core/graph/graph.cc [hunk 1]
```diff
@@ -1257,13 +1257,22 @@ void Graph::ReverseDFSFrom(const std::vector<NodeIndex>& from,
...
 
 void Graph::ReverseDFSFrom(const std::vector<const Node*>& from,
                            const std::function<void(const Node*)>& enter,
                            const std::function<void(const Node*)>& leave,
                            const std::function<bool(const Node*, const Node*)>& comp) const {
+
+  ReverseDFSFrom(from, enter, leave, comp, {});
+}
+
+void Graph::ReverseDFSFrom(const std::vector<const Node*>& from,
```
Comment (ytaous) [inline onnxruntime/core/graph/graph.cc:1271]: I saw a UT under ir/graph_test for ReverseDFS. Is it possible to add similar UT for the new method under orttraining?

Comment (suffiank) [reply onnxruntime/core/graph/graph.cc:1271]: Yes, will add similar UT.

File: orttraining/orttraining/core/session/training_session.h [hunk 2]
```diff
@@ -426,7 +426,11 @@ class TrainingSession : public InferenceSession {
                                       bool use_fp16_initializer,
                                       std::unordered_map<std::string, NodeArg*>& fp32_weight_name_to_fp16_node_arg);
 
-  std::unordered_set<std::string> GetTrainableModelInitializers(const ImmutableWeights& immutable_weights) const;
+  /** Discover all trainable initializers by reverse DFS from output
```
Comment (wschin) [inline orttraining/orttraining/core/session/training_session.h:429]:
```suggestion
  /** Discover all trainable initializers by reverse DFS starting from a given tensor (for example, the loss value)
```

File: orttraining/orttraining/core/session/training_session.cc [hunk 3]
```diff
@@ -188,11 +175,24 @@ Status TrainingSession::ConfigureForTraining(
...
 
+  // We need to get trainable weights to prevent constant folding from them. This works well if trainable weights are passed from config.
+  // For case we use GetTrainableModelInitializers to get trainable weights such as C++ frontend, it may get more initializers
+  // than trainable weights here as it's before transformers. So the constant folding may miss some nodes we actually can fold.
+  std::unordered_set<std::string> excluded_initializers =
+      !config.weight_names_to_train.empty()
+          ? config.weight_names_to_train
+          : GetTrainableModelInitializers(config.immutable_weights, loss_name);
+  for (const auto& weight_name_to_not_train : config.weight_names_to_not_train) {
+    excluded_initializers.erase(weight_name_to_not_train);
```
Comment (wschin) [inline orttraining/orttraining/core/session/training_session.cc:186]: Let's add
```
    LOGS(*session_logger_, INFO) << "Exclude "
                                 << weight_name_to_not_train
                                 << " from trainable list.";
```

File: orttraining/orttraining/core/session/training_session.cc [hunk 4]
```diff
@@ -188,11 +175,24 @@ Status TrainingSession::ConfigureForTraining(
...
+  ORT_RETURN_IF_ERROR(ApplyTransformationsToMainGraph(excluded_initializers));
+
   // derive actual set of weights to train
   std::unordered_set<std::string> weight_names_to_train =
       !config.weight_names_to_train.empty()
           ? config.weight_names_to_train
-          : GetTrainableModelInitializers(config.immutable_weights);
+          : GetTrainableModelInitializers(config.immutable_weights, loss_name);
   for (const auto& weight_name_to_not_train : config.weight_names_to_not_train) {
     weight_names_to_train.erase(weight_name_to_not_train);
```
Comment (wschin) [inline orttraining/orttraining/core/session/training_session.cc:197]: Similarly, can we add
```
Let's add
```
    LOGS(*session_logger_, INFO) << "Exclude "
                                 << weight_name_to_not_train
                                 << " from trainable list after applying graph transforms.";
```

File: include/onnxruntime/core/graph/graph.h [hunk 5]
```diff
@@ -737,6 +737,20 @@ class Graph {
                       const std::function<void(const Node*)>& leave,
                       const std::function<bool(const Node*, const Node*)>& comp = {}) const;
 
+    /** Performs a reverse depth-first search (DFS) traversal from a set of nodes, via their inputs,
```
Comment (KeDengMS) [inline include/onnxruntime/core/graph/graph.h:740]: >    [](start = 1, length = 3)

nit: spacing seems off

File: onnxruntime/core/graph/graph.cc [hunk 6]
```diff
@@ -1297,6 +1306,7 @@ void Graph::ReverseDFSFrom(const std::vector<const Node*>& from,
     if (comp) {
       std::vector<const Node*> sorted_nodes;
       for (auto iter = n.InputNodesBegin(); iter != n.InputNodesEnd(); ++iter) {
+        if (stop && stop(&n, &(*iter))) continue;
```
Comment (KeDengMS) [inline onnxruntime/core/graph/graph.cc:1309]: > if (stop && stop(&n, &(*iter))) continue; [](start = 7, length = 42)

can these two check on stop move to line 1301 and merge to one check? So basically we always push input nodes, but when visiting we'll skip when stop == true.

File: orttraining/orttraining/core/session/training_session.cc [hunk 7]
```diff
@@ -188,11 +175,24 @@ Status TrainingSession::ConfigureForTraining(
         config.model_with_loss_function_path.value(), SaveOption::NO_RELOAD));
   }
 
+  // We need to get trainable weights to prevent constant folding from them. This works well if trainable weights are passed from config.
+  // For case we use GetTrainableModelInitializers to get trainable weights such as C++ frontend, it may get more initializers
+  // than trainable weights here as it's before transformers. So the constant folding may miss some nodes we actually can fold.
+  std::unordered_set<std::string> excluded_initializers =
```
Comment (KeDengMS) [inline orttraining/orttraining/core/session/training_session.cc:181]: >excluded_initializers [](start = 34, length = 21)

would trainable_initializers a more descriptive variable name for this?

File: orttraining/orttraining/core/session/training_session.cc (hunk 4 above)
Comment (KeDengMS) [reply orttraining/orttraining/core/session/training_session.cc:197]: If there are many names, maybe some format like:
Excluding following weights from trainable list after applying graph transforms:
name0, name1, ...

---
In reply to: [434780997](https://github.com/microsoft/onnxruntime/pull/4116#discussion_r434780997) [](ancestors = 434780997)

File: orttraining/orttraining/core/session/training_session.cc [hunk 8]
```diff
@@ -937,26 +937,46 @@ bool TrainingSession::IsImmutableWeight(const ImmutableWeights& immutable_weight
...
-  std::unordered_set<std::string> trainable_initializers(model_initializers);
-  for (const std::string& initializer_name : model_initializers) {
-    const auto& nodes = graph.GetConsumerNodes(initializer_name);
-    for (const Node* node : nodes) {
-      if (IsUntrainable(node, initializer_name, session_logger_) ||
-          IsImmutableWeight(immutable_weights, node, initialized_tensors.at(initializer_name), session_logger_)) {
-        trainable_initializers.erase(initializer_name);
+  std::unordered_set<std::string> trainable_initializers;
+
+  auto add_valid_initializers = [&](const Node* node) {
```
Comment (KeDengMS) [inline orttraining/orttraining/core/session/training_session.cc:946]: >add_valid_initializers [](start = 7, length = 22)

add_trainable_initializers?

File: orttraining/orttraining/core/session/training_session.cc [hunk 9]
```diff
@@ -937,26 +937,46 @@ bool TrainingSession::IsImmutableWeight(const ImmutableWeights& immutable_weight
...
+            IsImmutableWeight(immutable_weights, node, initialized_tensors.at(initializer_name), session_logger_))
+          continue;
+
+        trainable_initializers.insert(initializer_name);
       }
-    }
-  }
+  };
+
+  auto stop_at_untrainable = [&](const Node* from, const Node* to) {
```
Comment (KeDengMS) [inline orttraining/orttraining/core/session/training_session.cc:960]: >stop_at_untrainable [](start = 7, length = 19)

stop_at_immutable_initializers

File: orttraining/orttraining/models/bert/main.cc [hunk 10]
```diff
@@ -554,7 +554,6 @@ void setup_training_params(BertParameters& params) {
   params.weights_not_to_train = {
       "position_01",            // Slice's dat input
       "op_min_ends_expand_10",  //op_min_ends_expand_10
```
Comment (KeDengMS) [inline orttraining/orttraining/models/bert/main.cc:556]: Are these still needed? I hope we could do this automatically

File: orttraining/orttraining/core/session/training_session.h [hunk 11]
```diff
@@ -426,7 +426,11 @@ class TrainingSession : public InferenceSession {
...
+  @param immutable_weights do not include initializers matching an (op_type, input_index, value) entry from this table
+  @param backprop_source_name reverse DFS back propagation source name (i.e. loss name or pipeline send output name)
+  */
+  std::unordered_set<std::string> GetTrainableModelInitializers(const ImmutableWeights& immutable_weights, const std::string& backprop_source_name) const;
```
Comment (xzhu1900) [inline orttraining/orttraining/core/session/training_session.h:433]: >std::unordered_set<std::string> GetTrainableModelInitializers(const ImmutableWeights& immutable_weights, const std::string& backprop_source_name) const; [](start = 2, length = 152)

nit: maximum line length is 120 (FYI onnx coding style: [https://github.com/microsoft/onnxruntime/blob/5e0928a777745a1536332973418dbbb7ccbf9d52/docs/Coding_Conventions_and_Standards.md](https://github.com/microsoft/onnxruntime/blob/5e0928a777745a1536332973418dbbb7ccbf9d52/docs/Coding_Conventions_and_Standards.md))

File: onnxruntime/core/graph/graph.cc (hunk 1 above)
Comment (ytaous) [reply onnxruntime/core/graph/graph.cc:1271]: I saw another place might be better for the UT - GradientGraphBuilderTest->TrainingSession_BertToy, perhaps duplicate that one and make your own if you like. 

---
In reply to: [434769892](https://github.com/microsoft/onnxruntime/pull/4116#discussion_r434769892) [](ancestors = 434769892)

File: onnxruntime/core/graph/graph.cc (hunk 6 above)
Comment (suffiank) [reply onnxruntime/core/graph/graph.cc:1309]: So the stop takes as input the current node (_from_ node) and the proposed neighbor to move to (_to_ node). Once the node is added to the stack, the preceding node it came from is lost. We can add that to the stack, but it seems simpler and more intuitive to me to perform the check beforehand. As a practical matter the number of calls will be the same.

File: onnxruntime/core/graph/graph.cc (hunk 1 above)
Comment (suffiank) [reply onnxruntime/core/graph/graph.cc:1271]: So I ended up modifying GraphTest.ReverseDFS. The _stop_ method modifies its traversal behavior.

File: onnxruntime/core/graph/graph.cc (hunk 6 above)
Comment (KeDengMS) [reply onnxruntime/core/graph/graph.cc:1309]: I see, thanks for the explanation, let's keep it this way.

---
In reply to: [436057581](https://github.com/microsoft/onnxruntime/pull/4116#discussion_r436057581) [](ancestors = 436057581)

File: orttraining/orttraining/models/bert/main.cc (hunk 10 above)
Comment (suffiank) [reply orttraining/orttraining/models/bert/main.cc]: It was previously needed to exclude these because there was no path to the loss node. This is now automatically detected and they are no longer needed.

Comment (KeDengMS) [reply orttraining/orttraining/models/bert/main.cc]: Awesome!
